core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1847,28 +1847,55 @@ protected Type visitSubqueryExpression(SubqueryExpression node, StackableAstVisi
                     .build();
             Scope queryScope = analyzer.analyze(node.getQuery(), subqueryScope);
 
-            // Subquery should only produce one column
-            if (queryScope.getRelationType().getVisibleFieldCount() != 1) {
-                throw semanticException(NOT_SUPPORTED,
-                        node,
-                        "Multiple columns returned by subquery are not yet supported. Found %s",
-                        queryScope.getRelationType().getVisibleFieldCount());
-            }
-
+            Type type;
             Node previousNode = context.getPreviousNode().orElse(null);
             if (previousNode instanceof InPredicate && ((InPredicate) previousNode).getValue() != node) {
+                // Subquery should only produce one column
+                if (queryScope.getRelationType().getVisibleFieldCount() != 1) {
+                    throw semanticException(NOT_SUPPORTED,
+                            node,
+                            "Multiple columns returned by subquery are not yet supported. Found %s",
+                            queryScope.getRelationType().getVisibleFieldCount());
+                }
+
                 subqueryInPredicates.add(NodeRef.of((InPredicate) previousNode));
+                type = getOnlyElement(queryScope.getRelationType().getVisibleFields()).getType();
             }
             else if (previousNode instanceof QuantifiedComparisonExpression) {
+                // Subquery should only produce one column
+                if (queryScope.getRelationType().getVisibleFieldCount() != 1) {
+                    throw semanticException(NOT_SUPPORTED,
+                            node,
+                            "Multiple columns returned by subquery are not yet supported. Found %s",
```

Comment:
> SELECT ROW(1, 2) < ALL (SELECT 3, 4)

I updated to handle this case, too

> SELECT ROW(1) < ALL (SELECT 3)?

Technically,  `1 < ALL (SELECT 3)` and `ROW(1) < ALL (SELECT 3)` are equivalent. The left side of those expressions is a `<row value expression>`, which is either `<row value special case>` or `<explicit row value constructor>`. The former handles the case of expressions that look like scalars but are to be interpreted as single-element row expressions.

This PR doesn't handle it. It will continue to fail as does today. We should tackled that as a separate improvement.